### PR TITLE
Correction in integration of two layers + correct start and end date including clear documentation

### LINKED ISSUE #155: Consistent start and end date yaml file
Is the begin and end date included or excluded in the tracking. This is different for the different timesteps in the yaml file.
Make this consistent (include all dates)

### LINKED ISSUE #177: Bug in main
We noted that there was a bug in the main version on 23-12-2022 which was related to the integration of the upper and lower layer in era5.py script. By looking into the history of the script era5.py we were able to detect where we made the change that led to the bug. The change in #119 was reversed regarding the integration of the two layers using model levels and pressure levels.

### PR DESCRIPTION
**Pull request by Imme1992**
What happened in this pull request.

Changed preprocess script and backtrack script such that the start and end date are included in preprocessing.
Changed backtrack script such that event tracking can be done per hour instead of per day.
Closes #155 

Fixed issue that the main was broken. Used history to reset to preprocess correctly (integration of two layers)
Fixes #177

### CODE REVIEW

**Comment by Imme1992** [comment]
issue #155 solved here

File: src/wam2layers/preprocessing/era5.py [hunk 1]
```diff
@@ -271,27 +269,43 @@ def prep_experiment(config_file):
...
-
+            #idx = dp.level.searchsorted(boundary, side="right")
+            #upper = np.s_[:, :idx, :, :]
+            #lower = np.s_[:, idx:, :, :]
+            #s_lower = cw[lower].sum(dim="level")
+            #s_upper = cw[upper].sum(dim="level")
+            #fx_lower = fx[lower].sum(dim="level")  # kg m-1 s-1
+            #fy_lower = fy[lower].sum(dim="level")  # kg m-1 s-1
+            #fx_upper = fx[upper].sum(dim="level")  # kg m-1 s-1
+            #fy_upper = fy[upper].sum(dim="level")  # kg m-1 s-1
```
Comment (Peter9192) [inline src/wam2layers/preprocessing/era5.py:280]: Please remove, don't comment out

File: scripts/download_era5_ml.py [hunk 2]
```diff
@@ -16,12 +16,12 @@
 import pandas as pd
```
Comment (ruudvdent) [inline scripts/download_era5_ml.py:16]: this file should not be in the PR

File: src/wam2layers/preprocessing/era5.py [hunk 3]
```diff
@@ -271,27 +269,43 @@ def prep_experiment(config_file):
...
-            fx_lower = fx.where(lower_layer).sum(dim="level")  # kg m-1 s-1
-            fy_lower = fy.where(lower_layer).sum(dim="level")  # kg m-1 s-1
-            fx_upper = fx.where(upper_layer).sum(dim="level")  # kg m-1 s-1
-            fy_upper = fy.where(upper_layer).sum(dim="level")  # kg m-1 s-1
+            #s_lower = cw.where(lower_layer).sum(dim="level")
+            #s_upper = cw.where(upper_layer).sum(dim="level")
+            #fx_lower = fx.where(lower_layer).sum(dim="level")  # kg m-1 s-1
+            #fy_lower = fy.where(lower_layer).sum(dim="level")  # kg m-1 s-1
+            #fx_upper = fx.where(upper_layer).sum(dim="level")  # kg m-1 s-1
+            #fy_upper = fy.where(upper_layer).sum(dim="level")  # kg m-1 s-1
```
Comment (Peter9192) [inline src/wam2layers/preprocessing/era5.py:294]: Same

File: src/wam2layers/config.py [hunk 4]
```diff
@@ -65,10 +65,11 @@ class Config(BaseModel):
...
 
-    preprocess_start_date: date
+    preprocess_start_date: datetime
     """Start date for preprocessing.
 
-    Should be formatted as: `"YYYY-MM-DD"`. Start date < end date.
+    Should be formatted as: `"YYYY-MM-DD[T]HH:MM"`. Start date < end date.
+    The preprocess_start_date is included in the preprocessing.
 
     For example:
```
Comment (ruudvdent) [inline src/wam2layers/config.py:74]: add example

File: src/wam2layers/tracking/backtrack.py [hunk 5]
```diff
@@ -485,12 +485,12 @@ def run_experiment(config_file):
...
 
-        # Only track the precipitation at certain dates
+        # Only track the precipitation at certain timesteps
         if (
             time_in_range(
                 config.event_start_date,
                 config.event_end_date,
-                t.date(),
+                t,
             )
```
Comment (Peter9192) [inline src/wam2layers/tracking/backtrack.py:494]: I was just wondering if this time_in_range function now also works without the 'strftime'

File: src/wam2layers/config.py [hunk 6]
```diff
@@ -90,11 +92,12 @@ class Config(BaseModel):
         preprocess_end_date: "2021-07-15"
```
Comment (ruudvdent) [inline src/wam2layers/config.py:92]: add example with T

File: scripts/download_era5_ml.py [hunk 7]
```diff
@@ -16,12 +16,12 @@
...
 from pathlib import Path
 
-target_dir = "."
+target_dir = "/projects/0/prjs0202/ERA5data"
```
Comment (Peter9192) [inline scripts/download_era5_ml.py:19]: Commits to download script accidentally committed?

File: src/wam2layers/config.py [hunk 8]
```diff
@@ -117,14 +121,15 @@ class Config(BaseModel):
         track_end_date: "2021-07-15"
```
Comment (ruudvdent) [inline src/wam2layers/config.py:121]: change to THH:MM

File: src/wam2layers/config.py [hunk 9]
```diff
@@ -135,14 +140,15 @@ class Config(BaseModel):
...
     start and end date of the event to something different than the total
-    tracking start and end date.
+    tracking start and end date, you can also indicate the hours that you want to track.
+    The event_end_date is included in the event tracking. 
 
-    Should be formatted as: `"YYYY-MM-DD"`. Start date < end date, even if
+    Should be formatted as: `"YYYY-MM-DD[T]HH:MM"`. Start date < end date, even if
     backtracking.
 
     For example:
```
Comment (ruudvdent) [inline src/wam2layers/config.py:154]: update example

**Review by Peter9192 — APPROVED** [review]
Thanks for the bugfix! Please edit the top post of the PR with more description, so that people looking back at this later will understand what was happening.

It would have been cleaner to make a separate PR for the bugfix and the updated functionality. But I like your pro-active attitude 😃

File: src/wam2layers/preprocessing/era5.py [hunk 10]
```diff
@@ -271,27 +269,43 @@ def prep_experiment(config_file):
...
-            upper = np.s_[:, :idx, :, :]
-            lower = np.s_[:, idx:, :, :]
-            s_lower = cw[lower].sum(dim="level")
-            s_upper = cw[upper].sum(dim="level")
-            fx_lower = fx[lower].sum(dim="level")  # kg m-1 s-1
-            fy_lower = fy[lower].sum(dim="level")  # kg m-1 s-1
-            fx_upper = fx[upper].sum(dim="level")  # kg m-1 s-1
-            fy_upper = fy[upper].sum(dim="level")  # kg m-1 s-1
-
+            #idx = dp.level.searchsorted(boundary, side="right")
```
Comment (ruudvdent) [inline src/wam2layers/preprocessing/era5.py:272]: should not be commented

File: src/wam2layers/config.py (hunk 9 above)
Comment (ruudvdent) [inline src/wam2layers/config.py:154]: add example

**Review by ruudvdent — CHANGES_REQUESTED** [review]
see comments

File: src/wam2layers/tracking/backtrack.py (hunk 5 above)
Comment (Imme1992) [reply src/wam2layers/tracking/backtrack.py:494]: Yes it works with strftime but it was not needed

File: src/wam2layers/config.py (hunk 9 above)
Comment (Imme1992) [reply src/wam2layers/config.py:154]: Done

File: src/wam2layers/preprocessing/era5.py (hunk 10 above)
Comment (Imme1992) [reply src/wam2layers/preprocessing/era5.py]: Done

File: src/wam2layers/config.py (hunk 9 above)
Comment (Imme1992) [reply src/wam2layers/config.py:154]: Done

**Review by ruudvdent — APPROVED** [review]

**Pull request merged**